livebooks/09_29_27_elixir_omt5.livemd

```diff
@@ -0,0 +1,186 @@
+query = from(u in User, where: u.age >= ^min_age and u.age <= ^max_age)
+Repo.all(query) |> Kino.DataTable.new()
+```
+
+## Checking for nil: the bad way
+
+![](https://i.giphy.com/media/4PCNGBPKj6De08Fgsc/giphy.webp)
+
+<!-- livebook:{"break_markdown":true} -->
+
+As soon as we call the function with nil, it will throw an error 😨
+
+```elixir
+defmodule BadExample do
+  import Ecto.Query
+
+  def by_age(min_age, max_age) do
+    query = from(u in User, where: u.age > ^min_age and u.age < ^max_age)
+    Repo.all(query) |> Kino.DataTable.new()
+  end
+end
+
+BadExample.by_age(21, nil) |> Kino.DataTable.new()
+# BadExample.by_age(nil, 55) |> Kino.DataTable.new()
+# BadExample.by_age(nil, nil) |> Kino.DataTable.new()
+```
+
+## Checking for nil: the ugly way
+
+![](https://i.giphy.com/media/Nxj5V2uURNzBt6TyWj/giphy.webp)
+
+<!-- livebook:{"break_markdown":true} -->
+
+As we have seen above, not checking for nil at all is a bad way to go. Let's check for nil for every possible case.
+
+* min_age = nil
+* max_age = nil
+* min_age = nil & max_age = nil
+
+```elixir
+defmodule UglyExample do
+  import Ecto.Query
+
+  def by_age(min_age, max_age) when is_nil(min_age) and is_nil(max_age) do
+    query = from(u in User)
+    Repo.all(query)
+  end
+
+  def by_age(min_age, max_age) when is_nil(min_age) do
+    query = from(u in User, where: u.age <= ^max_age)
+    Repo.all(query)
+  end
+
+  def by_age(min_age, max_age) when is_nil(max_age) do
+    query = from(u in User, where: u.age >= ^min_age)
+    Repo.all(query)
+  end
+
+  def by_age(min_age, max_age) do
+    query = from(u in User, where: u.age <= ^max_age and u.age > ^min_age)
+    Repo.all(query)
+  end
+end
+
+UglyExample.by_age(21, nil) |> Kino.DataTable.new()
+# UglyExample.by_age(nil, 55) |> Kino.DataTable.new()
+# UglyExample.by_age(nil, nil) |> Kino.DataTable.new()
+```
+
+This works fine, but it's a cumbersome and unelegant way. Elixir can do better, and we can, too 🤩
+
+## Checking for nil: the good way
+
+![](https://i.giphy.com/media/YaYUGMM8CYA02WW3je/giphy.webp)
+
+<!-- livebook:{"break_markdown":true} -->
+
+Here, we are implenting two helper functions:
+
+* maybe_add_min_age
+* maybe_add_max_age
+
+Inside those helper functions, we do our nil checks (loc: 13, loc: 16).
+
+Before we will do the actual query on the database, we pipe our query through those helper functions.
+
+```elixir
+defmodule GoodExample do
+  import Ecto.Query
+
+  def by_age(min_age, max_age) do
+    query = from(u in User)
+
+    query
+    |> maybe_add_min_age(min_age)
+    |> maybe_add_max_age(max_age)
+    |> Repo.all()
+  end
+
+  defp maybe_add_min_age(query, nil), do: query
+  defp maybe_add_min_age(query, min_age), do: where(query, [u], u.age >= ^min_age)
+
+  defp maybe_add_max_age(query, nil), do: query
+  defp maybe_add_max_age(query, max_age), do: where(query, [u], u.age <= ^max_age)
+end
+
+GoodExample.by_age(21, nil) |> Kino.DataTable.new()
+# GoodExample.by_age(nil, 55) |> Kino.DataTable.new()
```

Comment:
Why are these line commented out?

Reply:
These lines are commented just to call the function with  all possible way and for the users to play around.
What do you think? should we uncomment it or remove it. Its also in the uglyExample.